File: src/core/jupiter/core/common/sub/access/access-level.ts
import type { AccessStatus } from "@jupiter/webapi-client";
import { AccessLevel } from "@jupiter/webapi-client";

export function accessLevelName(accessLevel: AccessLevel): string {
  switch (accessLevel) {
    case AccessLevel.READER:
      return "Reader";
    case AccessLevel.COMMENTER:
      return "Commenter";
    case AccessLevel.WRITER:
      return "Writer";
    case AccessLevel.OWNER:
      return "Owner";
  }
}

export const SHARING_ACCESS_LEVELS: AccessLevel[] = [
  AccessLevel.READER,
  AccessLevel.COMMENTER,
  AccessLevel.WRITER,
];

export const ALL_ACCESS_LEVELS: AccessLevel[] = [
  AccessLevel.READER,
  AccessLevel.COMMENTER,
  AccessLevel.WRITER,
  AccessLevel.OWNER,
];

function accessLevelRank(accessLevel: AccessLevel): number {
  switch (accessLevel) {
    case AccessLevel.READER:
      return 0;
    case AccessLevel.COMMENTER:
      return 1;
    case AccessLevel.WRITER:
      return 2;
    case AccessLevel.OWNER:
      return 3;
  }
}

export function accessLevelAllows(
  accessLevel: AccessLevel,
  requiredAccessLevel: AccessLevel,
): boolean {
  return accessLevelRank(accessLevel) >= accessLevelRank(requiredAccessLevel);
}

export function accessStatusAllowsWriterOrAbove(
  accessStatus: AccessStatus | null | undefined,
): boolean {
  if (accessStatus === undefined || accessStatus === null) {
    return false;
  }

  return accessLevelAllows(accessStatus.access_level, AccessLevel.WRITER);
}

export function accessStatusIsOwner(
  accessStatus: AccessStatus | null | undefined,
): boolean {
  if (accessStatus === undefined || accessStatus === null) {
    return false;
  }

  return accessStatus.access_level === AccessLevel.OWNER;
}

File: src/core/jupiter/core/common/sub/access/access_level.py
"""The level of access a principal has to a shared resource."""

from jupiter.framework.value import EnumValue, enum_value


@enum_value
class AccessLevel(EnumValue):
    """The level of access a principal has to a shared resource."""

    OWNER = "owner"
    WRITER = "writer"
    COMMENTER = "commenter"
    READER = "reader"

    def allows(self, required: "AccessLevel") -> bool:
        """Whether this access level grants at least the required access level."""
        ranking = ["reader", "commenter", "writer", "owner"]
        return ranking.index(self.value) >= ranking.index(required.value)

    def is_strictly_stronger_than(self, other: "AccessLevel") -> bool:
        """Whether this access level is strictly stronger than another."""
        return self.allows(other) and self != other

    @property
    def is_invitable(self) -> bool:
        """Whether this level can be assigned when inviting a user to an entity."""
        return self != AccessLevel.OWNER
